//    Copyright 2018 Google Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        https://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

// This file defines extensions to FieldOptions and MessageOptions to enable
// annotation proto fields and messages with their corresponding FHIR types.

syntax = "proto3";

package google.fhir.stu3.proto;
option java_outer_classname = "Annotations";
option java_package = "com.google.fhir.stu3.proto";

import "google/protobuf/descriptor.proto";

// TODO: Unify with StructureDefinitionKindCode
enum StructureDefinitionKindValue {
  KIND_UNKNOWN = 0;
  KIND_PRIMITIVE_TYPE = 1;
  KIND_COMPLEX_TYPE = 2;
  KIND_RESOURCE = 3;
  KIND_LOGICAL = 4;
}

// To annotate cardinality constraints.
enum Requirement {
  NOT_REQUIRED = 0;
  REQUIRED_BY_FHIR = 1;
}

extend google.protobuf.MessageOptions {

  // If this message is a Code constrained to a specific valueset, this is the
  // valueset identifier.
  string fhir_valueset_url = 180887441;

  // If this message is a Reference, the reference is constrained to these
  // resource types.
  repeated string fhir_reference_type = 183546385;

  // What type of fhir structure does this message represent?
  StructureDefinitionKindValue structure_definition_kind = 182131192;

  // A brief description of the intended use of this message.
  string message_description = 190563156;

  // For primitive types, values must match this regex if present.
  string value_regex = 204543906;

  // Url for the structure definition this message was built from.
  string fhir_structure_definition_url = 207562726;

  // This element is a profile version of the base resource named here.
  string fhir_profile_base = 206336851;

  // DEPRECATED: use fhir_base_profile to identify extension, and
  // fhir_structure_definition_url to get the extension url.
  string fhir_extension_url = 177048773 [deprecated = true];
}

extend google.protobuf.EnumValueOptions {
  // If we had to rename this code to make a valid enum identifier, what was
  // the original name?
  string fhir_original_code = 181000551;
}

extend google.protobuf.FieldOptions {
  // Is this field required?
  Requirement validation_requirement = 162282766;

  // Is this field a wrapper around a choice type?
  bool is_choice_type = 187112991;

  // A brief description of the intended use of this field.
  string field_description = 190357977;

  // If this field has slices on it, this indicates the subfield to slice it by
  // (slicing.discriminator.path in the structure definition).
  string fhir_slice_by = 211855060;

  // This field is a slice of another field with this name.
  string fhir_slice_of = 211842556;

  // If this field is a slice, this is the value of the field indicated by the
  // fhir_slice_by_field on the main field that should be placed in this slice.
  string fhir_slice_value = 211855323;

  // Indicates the url for an extension slice.
  // This is a special case of the fhir_slice_value annotation that implies:
  // * fhir_slice_by = "url"  on the extension field
  // * fhir_slice_of = "extension" on this field
  // This has the additional behavior of inlining simple extensions (i.e.,
  // extensions with a value, rather than sub-extensions) as the value
  // itself.
  string fhir_inlined_extension_url = 206340037;

  // Indicates the system for a coding slice within a codeable concept.
  // This is a special case of the fhir_slice_value annotation that implies:
  // * fhir_slice_by = "code"  on the coding field
  // * fhir_slice_of = "coding" on this field
  // This has the additional behavior of inlining the code directly, instead
  // of a coding with fixed system.
  string fhir_inlined_coding_system = 211854975;

  // Indicates the code for a coding slice within a codeable concept.
  // This should only be used on fields that have fhir_inlined_coding_system
  // set.  In addition to the implications of fhir_inlined_coding_system,
  // this also sets a fixed code.
  string fhir_inlined_coding_code = 212122564;

  // Indicates which resource types can be valid references.
  repeated string valid_reference_type = 218884094;
}

extend google.protobuf.OneofOptions {
  // Whether or not this oneof is optional
  // If absent, oneofs are assumed to be REQUIRED if their containing message is
  // set. The most common case of oneofs is choice types, where the requirements
  // should be applied to the containing message.
  bool fhir_oneof_is_optional = 214472422;
}
